pages/api/notify-callback.ts

```diff
@@ -0,0 +1,101 @@
+ *
+ * Using the delivery status from the GC Nofity callback payload it will determine if a submission has to be reprocessed by the Reliability lambda.
+ *
+ * @returns
+ */
+async function processNotifyCallbackPayload(
+  req: NextApiRequest,
+  res: NextApiResponse
+): Promise<void> {
+  const submissionID = req.body.reference;
+  const deliveryStatus = req.body.status;
+
+  if (submissionID === undefined || deliveryStatus === undefined) {
+    return res
+      .status(400)
+      .json({ error: "Invalid payload: missing key attributes (reference or status)" });
+  }
+
+  if (deliveryStatus === "delivered" || deliveryStatus === "permanent-failure") {
+    return res.status(200).json({ status: "submission will not be reprocessed" });
+  }
+
+  const sqsClient = new SQSClient({
+    region: process.env.AWS_REGION ?? "ca-central-1",
+    endpoint: process.env.LOCAL_S3_ENDPOINT, // Using local S3 endpoint value as it is the same for SQS
+  });
+
+  try {
+    const getQueueURLCommand = new GetQueueUrlCommand({
+      QueueName: SQS_REPROCESS_SUBMISSION_QUEUE_NAME,
+    });
+
+    const getQueueURLCommandOutput = await sqsClient.send(getQueueURLCommand);
+
+    const sendMessageCommand = new SendMessageCommand({
+      MessageBody: JSON.stringify({ submissionID: submissionID }),
+      MessageDeduplicationId: submissionID,
+      MessageGroupId: "Group-" + submissionID,
+      QueueUrl: getQueueURLCommandOutput.QueueUrl,
+    });
+
+    await sqsClient.send(sendMessageCommand);
+
+    return res.status(200).json({ status: "submission will be reprocessed" });
+  } catch (error) {
+    logMessage.warn(error);
+    return res.status(500).json({ responses: "Processing error on the server side" });
+  }
+}
+
+export function validateAuthorizationHeader(
+  handler: (req: NextApiRequest, res: NextApiResponse) => void
+) {
+  return async function (req: NextApiRequest, res: NextApiResponse): Promise<unknown> {
+    try {
+      const bearerToken = extractBearerTokenFromReq(req);
+
+      if (bearerToken !== process.env.GC_NOTIFY_CALLBACK_BEARER_TOKEN) {
+        return res.status(403).json({ error: "Invalid authorization header" });
+      } else {
+        return handler(req, res);
+      }
+    } catch (error) {
+      return res.status(403).json({ error: "Missing authorization header" });
+    }
+  };
+}
+
+const isPostRequest = (handler: (req: NextApiRequest, res: NextApiResponse) => void) => {
+  return async (req: NextApiRequest, res: NextApiResponse): Promise<unknown> => {
+    try {
+      const requestBody = req.body && Object.keys(req.body).length > 0 ? req.body : undefined;
+      const method = requestBody?.method ? requestBody.method : req.method;
```

Comment:
FYI - @bryan-robitaille has a PR that is changing how `method` is being used. You'll need to update this to ensure it's in line with his changes, once they are merged.

Reply:
I will be looking into it for sure! There is something a little bit different and it is also the reason why I implemented my own functions. Since the API will be called by GC Notify we can't expect them to have an opened session so it becomes kind of a public and open API.
However, there is still some kind of protection because GC Notify will be passing us a bearer token that we can verify before processing any request.